### PR TITLE
xds: Add RLS in xDS e2e test

### PR DESCRIPTION
**Pull request by zasweq**
This PR adds an RLS in xDS e2e test. This tests that the RLS in xDS functionality works, and correctly configures and uses a RLS Load Balancer. This also pulls out some testing functionality from balancer/rls/internal to balancer/rls. This addresses issue #5216.

RELEASE NOTES: None

### CODE REVIEW

File: xds/internal/clusterspecifier/rls/rls.go [hunk 1]
```diff
@@ -40,6 +40,24 @@ func init() {
 	}
 }
 
+// RegisterForTesting registers the RLS Cluster Specifier Plugin for testing
+// purposes, regardless of the XDSRLS environment variable. This is needed
+// because there is no way to set the XDSRLS environment variable to true in a
+// test before init() in this package is run.
```
Comment (easwars) [inline xds/internal/clusterspecifier/rls/rls.go:46]: Does this mean that we can get rid of this once the env var is removed? If so, could you please add a TODO for the same here.

File: xds/internal/httpfilter/rbac/rbac.go [hunk 2]
```diff
@@ -43,16 +43,17 @@ func init() {
 	}
 }
 
-// RegisterForTesting registers the RBAC HTTP Filter for testing purposes, regardless
-// of the RBAC environment variable. This is needed because there is no way to set the RBAC
-// environment variable to true in a test before init() in this package is run.
+// RegisterForTesting registers the RBAC HTTP Filter for testing purposes,
+// regardless of the RBAC environment variable. This is needed because there is
+// no way to set the RBAC environment variable to true in a test before init()
+// in this package is run.
```
Comment (easwars) [inline xds/internal/httpfilter/rbac/rbac.go:49]: Same here about being able to remove this once the env var is gone. If so, please add TODO.

File: xds/internal/test/xds_client_integration_test.go [hunk 3]
```diff
@@ -28,12 +28,19 @@ import (
 	"testing"
 
 	"google.golang.org/grpc"
+	_ "google.golang.org/grpc/balancer/rls"
```
Comment (easwars) [inline xds/internal/test/xds_client_integration_test.go:31]: Please group blank imports in a separate group and add a one line comment for each as to why they are imported. I find that much more readable. Thanks.

File: xds/internal/test/xds_client_integration_test.go [hunk 4]
```diff
@@ -55,7 +62,7 @@ func clientSetup(t *testing.T, tss testpb.TestServiceServer) (uint32, func()) {
 	// Create a local listener and pass it to Serve().
 	lis, err := testutils.LocalTCPListener()
 	if err != nil {
-		t.Fatalf("testutils.LocalTCPListener() failed: %v", err)
+		t.Fatalf("test.LocalTCPListener() failed: %v", err)
```
Comment (easwars) [inline xds/internal/test/xds_client_integration_test.go:65]: Why do we need this change?

File: xds/internal/testutils/e2e/clientresources.go [hunk 5]
```diff
@@ -97,6 +98,21 @@ func DefaultClientResources(params ResourceParams) UpdateOptions {
 	}
 }
 
+// DefaultClientResourcesWithRLSCSP returns a set of resources (LDS, RDS, CDS, EDS) for a
+// client to connect to a server with a RLS Load Balancer as a child of Cluster Manager.
+func DefaultClientResourcesWithRLSCSP(params ResourceParams, rlsProto *rlspb.RouteLookupConfig) UpdateOptions {
```
Comment (easwars) [inline xds/internal/testutils/e2e/clientresources.go:103]: Does this function need to live here and be exported? Or can it live as an unexported function in the same package in which the test is defined. I don't foresee other tests needing this.

File: xds/internal/testutils/e2e/clientresources.go [hunk 6]
```diff
@@ -303,6 +319,33 @@ func DefaultRouteConfig(routeName, ldsTarget, clusterName string) *v3routepb.Rou
 	}
 }
 
+// DefaultRouteConfigWithRLSCSP returns a basic xds RouteConfig resource with an
+// RLS Cluster Specifier Plugin configured as the route.
+func DefaultRouteConfigWithRLSCSP(routeName, ldsTarget string, rlsProto *rlspb.RouteLookupConfig) *v3routepb.RouteConfiguration {
```
Comment (easwars) [inline xds/internal/testutils/e2e/clientresources.go:324]: Same here.

File: xds/internal/test/xds_client_integration_test.go [hunk 7]
```diff
@@ -246,3 +253,86 @@ func (s) TestClientSideRetry(t *testing.T) {
...
+// corresponding to this test service. This test asserts an RPC proceeds as normal
+// with the RLS Balancer as part of system.
+func (s) TestRLSinxDS(t *testing.T) {
+	oldRLS := envconfig.XDSRLS
+	envconfig.XDSRLS = true
+	defer func() {
+		envconfig.XDSRLS = oldRLS
+	}()
+	rlscsp.RegisterForTesting()
+	defer rlscsp.UnregisterForTesting()
```
Comment (easwars) [inline xds/internal/test/xds_client_integration_test.go:269]: Nit: combine the two statements being deferred into a single defer call?

File: xds/internal/test/xds_client_integration_test.go [hunk 8]
```diff
@@ -246,3 +253,86 @@ func (s) TestClientSideRetry(t *testing.T) {
...
+	if _, err := client.EmptyCall(ctx, &testpb.Empty{}, grpc.WaitForReady(true)); err != nil {
+		t.Fatalf("rpc EmptyCall() failed: %v", err)
+	}
+
+	// These RLS Verifications makes sure the RLS Load Balancer is actually part
+	// of the xDS Configured system that correctly sends out RPC.
+
+	// Verify connection is established to RLS Server.
+	_, err = lis.NewConnCh.Receive(ctx)
+	if err != nil {
```
Comment (easwars) [inline xds/internal/test/xds_client_integration_test.go:328]: Please combine these two lines.

**Review by easwars — CHANGES_REQUESTED** [review]
Wrt to moving code around from `balancer/rls/helpers_test.go` and `balancer/rls/internal/test/e2e/rls_fakeserver.go` to `balancer/rls/test/rls.go`, it seems to be moving stuff out of an `internal` directory into a non-internal directory.

May I suggest the following:
- Types like `ListenerWrapper` and `ConnWrapper` are nothing specific to rls. You might consider moving them to `internal/testutils` package
- Other functionality like spinning up the fake RLS server etc, you might consider moving to some common, but internal directory. Maybe something like `internal/testutils/rls` or `internal/rls/testutils`. The only downside of these two approaches is that it would require import renaming when they are used. But I would prefer that instead of having these in non-internal directories.

**Review by zasweq — COMMENTED** [review]
Thanks for the comments :D! I originally moved testing helpers to internal/testutils, but there was a cyclic dependency with grpc (the fake RLS Server uses grpc). Your suggestion of moving the Conn/Listener wrappers to internal/testutils and moving the fake RLS Server to internal/testutils/rls fixed this :D!

File: balancer/rls/helpers_test.go [hunk 9]
```diff
@@ -62,51 +60,8 @@ func Test(t *testing.T) {
...
-		}
-	}
-
-	return &listenerWrapper{
-		Listener:  lis,
-		newConnCh: testutils.NewChannel(),
-	}
-}
+// Move Listener Wrapper, newListenerWrapper, and setupFakeRLSServer to rls_fakeserver.go
+// 1. Delete what was present here that I moved to prevent code duplication - then fix balancer test
```
Comment (easwars) [inline balancer/rls/helpers_test.go:64]: Please remove this comment.

File: balancer/rls/picker_test.go [hunk 10]
```diff
@@ -143,8 +143,8 @@ func (s) TestPick_DataCacheMiss_PendingEntryExists(t *testing.T) {
 		},
 	}
 
-	for _, test := range tests {
-		t.Run(test.name, func(t *testing.T) {
+	for _, tst := range tests {
```
Comment (easwars) [inline balancer/rls/picker_test.go:146]: No need to rename `test` to `tst` now that the testing package is called `rlstest`.

File: balancer/rls/picker_test.go [hunk 11]
```diff
@@ -261,12 +261,12 @@ func (s) TestPick_DataCacheHit_NoPendingEntry_StaleEntry(t *testing.T) {
...
```
Comment (easwars) [inline balancer/rls/picker_test.go:264]: Same here. Undo renaming.

File: balancer/rls/picker_test.go [hunk 12]
```diff
@@ -361,12 +361,12 @@ func (s) TestPick_DataCacheHit_NoPendingEntry_ExpiredEntry(t *testing.T) {
...
```
Comment (easwars) [inline balancer/rls/picker_test.go:364]: And here. And in all tests below.

File: internal/testutils/rls/fake_rls_server.go [hunk 13]
```diff
@@ -16,7 +16,8 @@
  *
  */
 
-package e2e
+// Package rls contains utilities for RouteLookupService tests.
```
Comment (easwars) [inline internal/testutils/rls/fake_rls_server.go:19]: Package rls contains utilities for RouteLookupService e2e tests?

File: internal/testutils/rls/fake_rls_server.go [hunk 14]
```diff
@@ -86,8 +110,8 @@ func (s *FakeRouteLookupServer) RouteLookup(ctx context.Context, req *rlspb.Rout
 	if s.respCb == nil {
 		return &rlspb.RouteLookupResponse{}, nil
 	}
-	resp := s.respCb(ctx, req)
-	return resp.Resp, resp.Err
+	resp := s.respCb(ctx, req) // func(context.Context, *rlspb.RouteLookupRequest) *RouteLookupResponse - write your own function that sends correct CDS response backward
+	return resp.Resp, resp.Err // Set this target to correspond to correct cluster for RLS to proceed as normal
```
Comment (easwars) [inline internal/testutils/rls/fake_rls_server.go:114]: Are these changes required?

File: xds/internal/test/xds_client_integration_test.go [hunk 15]
```diff
@@ -30,11 +30,25 @@ import (
...
+	rlstest "google.golang.org/grpc/internal/testutils/rls"
 	"google.golang.org/grpc/status"
+	rlscsp "google.golang.org/grpc/xds/internal/clusterspecifier/rls"
 	"google.golang.org/grpc/xds/internal/testutils/e2e"
+	"google.golang.org/protobuf/types/known/durationpb"
 
+	// To register the RLS Load Balancing policy.
+	_ "google.golang.org/grpc/balancer/rls"
+	// To register the RLS Cluster Specifier Plugin.
+	_ "google.golang.org/grpc/xds/internal/clusterspecifier/rls"
```
Comment (easwars) [inline xds/internal/test/xds_client_integration_test.go:46]: I meant something like this:

```golang
_ "google.golang.org/grpc/balancer/rls" // Register the RLS Load Balancing policy.
_ "google.golang.org/grpc/xds/internal/clusterspecifier/rls" // Register the RLS Cluster Specifier Plugin.
```

Sorry for the nit.

**Review by easwars — CHANGES_REQUESTED** [review]

**Comment by easwars** [comment]
In https://github.com/grpc/grpc-go/pull/5285, I figured out a way to use the `stubBalancer` to use a real balancer underneath it. Once that is merged, we could get rid of the code in `rls_child_policy.go` and use the former instead.

**Review by zasweq — COMMENTED** [review]
Thanks for the comments :D! Sorry about all the personal checked in comments. Noted @ your stub balancer comment.

**Review by easwars — APPROVED** [review]

**Pull request merged**